radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/data/TransactionAction.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * (C) Copyright 2021 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.api.data;
+
+import com.radixdlt.atom.TxAction;
+import com.radixdlt.atom.actions.BurnToken;
+import com.radixdlt.atom.actions.CreateFixedToken;
+import com.radixdlt.atom.actions.CreateMutableToken;
+import com.radixdlt.atom.actions.MintToken;
+import com.radixdlt.atom.actions.RegisterValidator;
+import com.radixdlt.atom.actions.StakeTokens;
+import com.radixdlt.atom.actions.TransferToken;
+import com.radixdlt.atom.actions.UnregisterValidator;
+import com.radixdlt.atom.actions.UnstakeTokens;
+import com.radixdlt.atom.actions.UpdateValidator;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.identifiers.REAddr;
+import com.radixdlt.utils.UInt256;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static java.util.Optional.ofNullable;
+
+public class TransactionAction {
+	private final ActionType actionType;
+	private final REAddr from;
+	private final REAddr to;
+	private final ECPublicKey delegate;
+	private final UInt256 amount;
+	private final REAddr rri;
+	private final String name;
+	private final String url;
+	private final String symbol;
+	private final String iconUrl;
+	private final String tokenUrl;
+	private final String description;
+
+	private TransactionAction(
+		ActionType actionType,
+		REAddr from,
+		REAddr to,
+		ECPublicKey delegate,
+		UInt256 amount,
+		REAddr rri,
+		String name,
+		String url,
+		String symbol,
+		String iconUrl,
+		String tokenUrl,
+		String description
+	) {
+		this.actionType = actionType;
+		this.from = from;
+		this.to = to;
+		this.amount = amount;
+		this.delegate = delegate;
+		this.rri = rri;
+		this.name = name;
+		this.url = url;
+		this.symbol = symbol;
+		this.iconUrl = iconUrl;
+		this.tokenUrl = tokenUrl;
+		this.description = description;
+	}
```

Comment:
Some idea: as most of the fields are optional, this would be a good candidate for the builder design pattern.

Reply:
Good point. I'll take a look if it worth the hassle.

Reply:
Deeper look shows that all these methods are accessed via method reference. Replacing them with builder pattern will make code unnecessarily verbose with no obvious benefits